29.md

```diff
@@ -0,0 +1,67 @@
+NIP-29
+======
+
+Time-Based Sync
+---------------
+
+`draft` `optional` `author:vitorpamplona`
+
+This NIP describes a simple event database reconciliation procedure for clients and relays (peers). Both sides hash the same groups of event ids and compare the resulting hashes to determine which subsets should be downloaded or uploaded to each other. The procedure is ideal for clients that use a local database and must guarantee the database is in sync with each relay without downloading or uploading all events again. 
+
+### Sync Protocol
+
+The syncing peer (client or a relay) sends a `HASH-REQ` message to a relay with a subscription ID and filters for the content to be synced. 
+
+Request:
+```js
+[
+    "HASH-REQ",
+    "<subscription ID string>",
+    "<WINDOW-SIZE>"	
+    <nostr filter>, <nostr filter2>, <nostr filter3>
+]
+```
+
+Upon receiving a `HASH-REQ` the relay MUST: 
+
+1. Apply all nostr filters in the subscription to the database
+2. Sort all resulting events by `.created_at`
+3. Group by the first `WINDOW-SIZE` chars of the stringified `.created_at`
+4. For each group, create an array of event ids: `["idHex1", "idHex2", "idHex3"]`, JSON-serialize it, and hash it using SHA-256
```

Comment:
Should probably specify that the ids are sorted

Reply:
good catch